_drafts/debug-adapter-rundown.md

```diff
@@ -0,0 +1,215 @@
+
+![](/img/debug-adapter-rundown/stlink-v2.jpg){:height="250px" style="float:
+left"} ![](/img/debug-adapter-rundown/stlink-v3mini.jpg){:height="250px"}
+
+ST Microelectronics' "ST-Link" debug probe has been around for a long time. It's
+commonly included on dev boards as a built-in debug probe.
+
+The probe devices + on-board software, as well as ST's default host-side tools,
+are all closed source, but there are several open-source tools that are able to
+talk to the ST-Link devices (openocd, PyOCD, probe-rs, etc).
+
+There's several generations of the ST-Link probe. The V2(.1) are probably the
+most common.
+
+- overview of ST-Link varieties:
+  <https://www.st.com/resource/en/technical_note/tn1235-overview-of-stlink-derivatives-stmicroelectronics.pdf>
+- open source tools: <https://github.com/stlink-org/stlink>
+
+### Black Magic Probe
+
+![](/img/debug-adapter-rundown/black-magic.png){:height="250px"}
+
+The Black Magic Probe is a fully open source (hardware + software) debug probe.
+One standout feature of this device is the gdb server runs on the probe itself;
+it appears as a serial device when plugged into the host computer, and you can
+directly target it with `extended-remote` without any other software required.
+
+The Black Magic Probe firmware is also capable of being installed on a few other
+hardware devices, including ST-Link's and stm32 boards:
+
+<https://github.com/blackmagic-debug/blackmagic/tree/e82d4f2edaefb74c9dee27f6d26c4d2038af0384/src/platforms>
+
+See the links below for some more information:
+
+- <https://1bitsquared.com/products/black-magic-probe>
+- <https://github.com/blackmagic-debug/blackmagic/wiki>
+
+### DAPLink
+
+![](/img/debug-adapter-rundown/MAX32625PICO.png){:height="250px"}
+
+DAPLink is an open-source intiative started by ARM Mbed. It defines a standard
+interface for host tools to interact with any Cortex-M target via a USB HID
+based adapter device (the probe itself).
+
+The original DAPLink probes are no longer stocked, but a de-facto replacement in
+the [MAX32625PICO](https://os.mbed.com/platforms/MAX32625PICO).
+
+Also, many arduino devices can operate as DAPLink adapters, for example the very
+inexpensive "XIAO" SAMD21 dev board:
+
+<https://www.seeedstudio.com/Seeeduino-XIAO-Arduino-Microcontroller-SAMD21-Cortex-M0+-p-4426.html>
+
+There's also a WebUSB-based host available:
+
+<https://github.com/ARMmbed/dapjs>
+
+Some more information in the following links:
+
+- <https://github.com/armmbed/DAPLink>
+- <https://github.com/Seeed-Studio/Seeed_Arduino_DAPLink/>
+
+### FTDI FT2232H
+
+The classic FTDI Serial-to-USB adapter IC's can be used as generic SWD adapters.
+
+OpenOCD has a built-in configuration for using those devices as a debug probe:
+
+<https://openocd.org/doc/html/Debug-Adapter-Hardware.html>
+
+Depending on which FTDI chip you're using, you can use spare UARTs for a debug
+console to your target.
+
+One nice thing about this approach is it's possible to add the FTDI chip to your
```

Comment:
I'd argue that you can do this with DAPLink as well

Reply:
Ohh good point 👌